crates/octez/src/async/client.rs

```diff
@@ -331,6 +332,81 @@ impl OctezClient {
         self.spawn_and_wait_command(args).await?;
         Ok(())
     }
+
+    pub async fn originate_contract(
+        &self,
+        contract_alias: &str,
+        funding_account_src: &str, // can be an address or an alias
+        fund_tez: f64,
+        contract_path: &Path,
+        init_data: Option<&str>,
+        burn_cap: Option<f64>,
+    ) -> Result<(ContractKt1Hash, OperationHash)> {
+        let fund_str = fund_tez.to_string();
+        let burn_cap_str = match burn_cap {
+            Some(v) => v.to_string(),
+            None => String::new(),
+        };
+        let mut args = vec![
+            "originate",
+            "contract",
+            contract_alias,
+            "transferring",
+            &fund_str,
+            "from",
+            funding_account_src,
+            "running",
+            contract_path
+                .to_str()
+                .ok_or(anyhow::anyhow!("failed to convert contract path to string"))?,
+        ];
+        if let Some(data) = init_data {
+            args.append(&mut vec!["--init", data]);
+        }
+        if !burn_cap_str.is_empty() {
+            args.append(&mut vec!["--burn-cap", &burn_cap_str]);
+        }
+        let output = self.spawn_and_wait_command(args).await?;
+
+        let operation_hash = parse_operation_hash(&output);
+        if operation_hash.is_none() {
+            return Err(anyhow::anyhow!(
+                "failed to parse operation hash from execution output"
+            ));
+        }
+        let contract_address = parse_contract_address(&output);
+        if contract_address.is_none() {
+            return Err(anyhow::anyhow!(
+                "failed to parse contract address from execution output"
+            ));
+        }
+        Ok((contract_address.unwrap(), operation_hash.unwrap()))
+    }
+}
+
+fn parse_operation_hash(output: &str) -> Option<OperationHash> {
+    let operation_hash_pattern =
+        regex::Regex::new("Operation hash is '(o[1-9A-HJ-NP-Za-km-z]{50})'").unwrap();
+    if let Some(capture) = operation_hash_pattern.captures(output) {
+        let raw_operation_hash = capture.get(1).unwrap().as_str();
+        if let Ok(v) = OperationHash::from_base58_check(raw_operation_hash) {
+            return Some(v);
+        }
+    }
+    None
+}
+
+fn parse_contract_address(output: &str) -> Option<ContractKt1Hash> {
```

Comment:
Seems like a lot of repetition between `parse_contract_address` and `parse_operation_hash`. You can factorize it into:
 1. Extract regex
 2. Parse as hash

Reply:
I wouldn't call this "a lot of repetition" and extracting the common parts does not make things much more readable, but I've updated it anyway.

Reply:
the client also has a logic that parses the tez balance printed out by the client as well(e.g. 100 ꜩ)
 What about we create a struct for parsing client stdout that contains all these logic ?

Reply:
Yup. That would be in [JSTZ-132](https://linear.app/tezos/issue/JSTZ-132/implement-ascommand-trait-helpers) and [JSTZ-133](https://linear.app/tezos/issue/JSTZ-133/implement-ascommand-for-octezclient).